From f99d195730635b0d2a649adf1087c0ecf67ec0da Mon Sep 17 00:00:00 2001
From: Kathiravan T <quic_kathirav@quicinc.com>
Date: Wed, 17 May 2023 11:55:20 +0530
Subject: [PATCH 248/281] firmware: qcom_scm: use SMC_CONVENTION_ARM_32
 convention for 32bit kernel

During SCM probe, to identify the SCM convention, first SCM call is made
with SMC_CONVENTION_ARM_64. However this is causing crash on 32bit
kernel, due to changes in the register sets between ARM and AARCH64.
To avoid this, use SMC_CONVENTION_ARM_32 if its 32bit kernel.

Change-Id: I84eefd91a00e303b80388e43f59dc5c914352bca
Signed-off-by: Kathiravan T <quic_kathirav@quicinc.com>
---
 drivers/firmware/qcom_scm.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/firmware/qcom_scm.c b/drivers/firmware/qcom_scm.c
index 523cd95e9518..f495af02a6e1 100644
--- a/drivers/firmware/qcom_scm.c
+++ b/drivers/firmware/qcom_scm.c
@@ -164,6 +164,13 @@ static enum qcom_scm_convention __get_convention(void)
 	if (likely(qcom_scm_convention != SMC_CONVENTION_UNKNOWN))
 		return qcom_scm_convention;
 
+	/*
+	 * When running on 32bit kernel, SCM call with convention
+	 * SMC_CONVENTION_ARM_64 is causing the system crash. To avoid that
+	 * use SMC_CONVENTION_ARM_64 for 64bit kernel and SMC_CONVENTION_ARM_32
+	 * for 32bit kernel.
+	 */
+#if IS_ENABLED(CONFIG_ARM64)
 	/*
 	 * Device isn't required as there is only one argument - no device
 	 * needed to dma_map_single to secure world
@@ -184,6 +191,7 @@ static enum qcom_scm_convention __get_convention(void)
 		forced = true;
 		goto found;
 	}
+#endif
 
 	probed_convention = SMC_CONVENTION_ARM_32;
 	ret = __scm_smc_call(NULL, &desc, probed_convention, &res, true);
-- 
2.17.1

